internal/gengapic/genrest.go

```diff
@@ -210,33 +210,94 @@ func (g *generator) queryParams(m *descriptor.MethodDescriptorProto) map[string]
 	// This may be the request message itself or a subfield.
 	// Body parameters are not valid query parameters,
 	// because that means the same param would be sent more than once.
-	var bodyMsg *descriptor.DescriptorProto
 	bodyField := g.lookupField(m.GetInputType(), info.body)
-	if bodyField.GetType() == descriptor.FieldDescriptorProto_TYPE_MESSAGE {
-		bodyMsg = g.descInfo.Type[bodyField.GetTypeName()].(*descriptor.DescriptorProto)
-	}
 
-	for _, f := range request.GetField() {
-		// If, and only if, a field is not a path parameter or a body parameter,
+	// Possible query parameters are all leaf fields in the request or body.
+	pathToLeaf := g.getLeafs(request, bodyField)
+	// Iterate in sorted order to
+	for path, leaf := range pathToLeaf {
+		// If, and only if, a leaf field is not a path parameter or a body parameter,
 		// it is a query parameter.
-		if _, ok := pathParams[f.GetName()]; !ok && g.lookupField(bodyMsg.GetName(), f.GetName()) == nil {
-			queryParams[f.GetName()] = f
+		if _, ok := pathParams[path]; !ok && g.lookupField(request.GetName(), leaf.GetName()) == nil {
+			queryParams[path] = leaf
 		}
 	}
 
 	return queryParams
 }
 
+// Returns a map from fully qualified path to field descriptor for all the leaf fields of a message 'm',
+// where a "leaf" field is a non-message whose top message ancestor is 'm'.
+// e.g. for a message like the following
+//
+// message Mollusc {
+//     message Squid {
+//         message Mantle {
+//             int32 mass_kg = 1;
+//         }
+//         Mantle mantle = 1;
+//     }
+//     Squid squid = 1;
+// }
+//
+// The one entry would be
+// "squid.mantle.mass_kg": *descriptor.FieldDescriptorProto...
+func (g *generator) getLeafs(msg *descriptor.DescriptorProto, excludedFields ...*descriptor.FieldDescriptorProto) map[string]*descriptor.FieldDescriptorProto {
```

Comment:
Why is `excludedFields` variadic when the only call site provides one item? Seems like it doesn't need to be variadic.

Reply:
This is a combination of laziness and future proofing on my part. I want to be able to exclude the `body` field when calling `getLeafs` from `queryParams`, but I didn't want to rewrite the tests to support the parameter. Making `excludedFields` variadic is a cheater way of making it optional.